### ISSUE TITLE
Disallow buttons with the same color as the background color

### ISSUE THREAD

**Issue by dance2die** [forem/forem#735]
## Feature Request or Task
Would you check when the `follow` button is the same color as background thus hiding the button?

As an example, 
Pierre Baron on https://dev.to/prbaron has following profile page,
making it hard to follow.

![image](https://user-images.githubusercontent.com/8465237/46016787-b8f3fc80-c0a3-11e8-9801-0807588ffb69.png)

__________

And the button is actually there.
![image](https://user-images.githubusercontent.com/8465237/46016751-96fa7a00-c0a3-11e8-9a7e-72bebdfcd492.png)

**Comment by Zhao-Andy** [comment]
We have a bit of a check that does this, but it's imperfect. Thanks for the heads up!

**Comment by jessleenyc** [comment]
What @Zhao-Andy said -- we take care of this in some places where the background color and color are the _exact same_, which still isn't ideal for colors that are very similar but not identical. 

Would be great if we could add a minimum contrast requirement to buttons.

**Comment by hdennen** [comment]
Seems simpler to render the button in the settings page to show the changes the user makes. Maybe immediate feedback > safety code in this case. ¯\_(ツ)_/¯

**Comment by Link2Twenty** [comment]
It may be worth checking the contrast and if it fails the aria spec setting the text to solid white or black, which would be calculated at the time.

https://www.w3.org/TR/WCAG20/#visual-audio-contrast
https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef

I've thrown together a quick proof of concept, it's written in JS but I imagine you'd rather the backend handled it.

http://jsfiddle.net/link2twenty/jL52defu/

**Comment by Link2Twenty** [comment]
I've not ever written ruby before so feel free to write something yourself but something like this is what I was thinking

```ruby
def contrast_verified_hex(foreground, background)
  fg = foreground[1..-1].chars.each_slice(2).map(&:join)
  bg = background[1..-1].chars.each_slice(2).map(&:join)
  
  f = {
    R: fg[0].hex / 255.0,
    G: fg[1].hex / 255.0,
    B: fg[2].hex / 255.0
  }
  
  b = {
    R: bg[0].hex / 255.0,
    G: bg[1].hex / 255.0,
    B: bg[2].hex / 255.0
  }
  
  f_R_L = f[:R] <= 0.03928 ? (f[:R] * (1 / 12.92)) : (((f[:R] + 0.055) / 1.055) ** 2.4)
  f_G_L = f[:G] <= 0.03928 ? (f[:G] * (1 / 12.92)) : (((f[:G] + 0.055) / 1.055) ** 2.4)
  f_B_L = f[:B] <= 0.03928 ? (f[:B] * (1 / 12.92)) : (((f[:B] + 0.055) / 1.055) ** 2.4)
  
  b_R_L = b[:R] <= 0.03928 ? (b[:R] * (1 / 12.92)) : (((b[:R] + 0.055) / 1.055) ** 2.4)
  b_G_L = b[:G] <= 0.03928 ? (b[:G] * (1 / 12.92)) : (((b[:G] + 0.055) / 1.055) ** 2.4)
  b_B_L = b[:B] <= 0.03928 ? (b[:B] * (1 / 12.92)) : (((b[:B] + 0.055) / 1.055) ** 2.4)
  
  fg_lum = (0.2126 * f_R_L + 0.7152 * f_G_L + 0.0722 * f_B_L) + 0.05
  bg_lum = (0.2126 * b_R_L + 0.7152 * b_G_L + 0.0722 * b_B_L) + 0.05
  
  ratio = fg_lum / bg_lum
  
  if fg_lum < bg_lum
    ratio = 1 / ratio
  end
  
  if ratio >= 7
    return foreground
  end
  
  if bg_lum < 0.5
    return "#FFFFFF"
  end
  
  return "#000000"
end
```

Then in `user_decorator.rb` you could say something like

``` ruby
def enriched_colors
    if bg_color_hex.blank?
      {
        bg: assigned_color[:bg],
        text: assigned_color[:text]
      }
    else
      {
        bg: bg_color_hex || assigned_color[:bg],
        text: contrast_verified_hex(text_color_hex, bg_color_hex) || assigned_color[:text]
      }
    end
  end
```

**Comment by maestromac** [comment]
Thanks for sharing those codes @Link2Twenty !

**Comment by Link2Twenty** [comment]
Is there anything I can do to move this forward? Or are we waiting for a decision of some sort?

**Comment by jessleenyc** [comment]
@Link2Twenty thanks for the ping. This is approved if you'd like to work on it!

**Comment by Link2Twenty** [comment]
How are we hoping to deal with it? Show the user their current ratio, perhaps just 'good' or 'bad'. Or would we rather just override the users colour?

The former is politer but would be harder to implement.

**Comment by jessleenyc** [comment]
Hm, perhaps override the user colors but have a message appear that says the colors have been adjusted to meet requirements? Thoughts on this @maestromac?

**Comment by maestromac** [comment]
Yeah that's a good idea @jessleenyc .

**Comment by benhalpern** [comment]
Yeah @jessleenyc that makes sense. Are we talking about the message being displayed on the profile settings?

**Comment by jessleenyc** [comment]
@benhalpern yep!

**Comment by Link2Twenty** [comment]
Now that we have a preview panel it might be nice to have some sort of contrast indicator.
Ref: #4212

**Comment by Link2Twenty** [comment]
I was having a little think about how the UI should look for this. I settled on something like this.

![mock up](https://user-images.githubusercontent.com/3534427/66649255-9b012780-ec25-11e9-9c27-5f517f983e56.png)

I've been updating my [fiddle](https://jsfiddle.net/link2twenty/zt58yhuj/) in the above comment for if anyone wants to look at this.

**Comment by cmgorton** [comment]
Hey all. Thanks for bringing this up. This issue will now be encompassed in the broader accessibility issue Marcy has opened #11504 . If we have any more comments or suggestions we can leave them in that issue. 
Since that is the case I am going to close this issue.
